src/mlpack/methods/dbscan/dbscan_impl.hpp

```diff
@@ -164,53 +184,62 @@ void DBSCAN<RangeSearchType, PointSelectionPolicy>::PointwiseCluster(
   // We maintain a list of non-core points so that we can handle that logic
   // correctly.
 
-  std::vector<bool> visited(data.n_cols, false);
-  std::vector<bool> nonCorePoints(data.n_cols, false);
+  std::vector<std::atomic<bool>> visited(data.n_cols);
+  std::vector<std::atomic<bool>> nonCorePoints(data.n_cols);
 
   for (size_t i = 0; i < data.n_cols; ++i)
   {
-    if (i % 10000 == 0 && i > 0)
```

Comment:
I believe that all of the optimizations below here are invalid and cause this algorithm to be something other than DBSCAN. I hate to say that because these give nice speedups (and are the core of the speedup of this PR) but inhere tly DBSCAN is a serial algorithm: select a point, find other points within range, mark them as part of the same cluster, repeat.  That can't be parallelized (or I mean you can, but then it's not DBSCAN so we can't do it).